File: csharp/Practice01/src/MyClient/MyReceiver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections.Concurrent;
using System.Threading;

namespace MyClient
{
    public sealed class MyReceiver
    {
        private readonly Random _random = new Random(DateTime.Now.Millisecond);

        private readonly HashSet<int> _queryIds = new HashSet<int>();
        private readonly BlockingCollection<MyData> _dataCollection = new BlockingCollection<MyData>(10000);
        private readonly CancellationTokenSource _cts = new CancellationTokenSource();

        internal MyReceiver()
        {
            new Thread(FeedData).Start();
        }

        internal void AddQuery(int id)
        {
            lock (this._queryIds)
            {
                this._queryIds.Add(id);
                this._dataCollection.Add(new MyData(id, "begin"));
            }
        }

        internal void RemoveQuery(int id)
        {
            lock (this._queryIds)
            {
                this._queryIds.Remove(id);
            }
        }

        internal void Close()
        {
            this._cts.Cancel();
        }

        private void FeedData()
        {
            while (!this._cts.IsCancellationRequested)
            {
                Thread.Sleep(500);

                lock (this._queryIds)
                {
                    foreach (var id in this._queryIds)
                    {
                        this._dataCollection.Add(new MyData(id, this._random.Next().ToString()));
                    }
                }
            }
        }

        public MyData Receive()
        {
            if (this._cts.IsCancellationRequested) return null;

            if (this._random.NextDouble() < 0.01)
            {
                throw new MyClientException("Error occurred when receive.");
            }

            try
            {
                return this._dataCollection.Take(this._cts.Token);
            }
            catch (OperationCanceledException)
            {
                return null;
            }
        }
    }
}

File: java/Practice01/src/myClient/MyReceiver.java
package myClient;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;

public final class MyReceiver {
	
	private Set<Integer> _queryIds = new HashSet<Integer>();
	private BlockingQueue<MyData> _dataQueue = new LinkedBlockingQueue<MyData>();
	private AtomicBoolean _isClosed = new AtomicBoolean(false);
	
	protected MyReceiver() {
		new Thread(new Runnable() {
			
			@Override
			public void run() {
				MyReceiver.this.feedData();
			}
			
		}).start();
	}
	
	protected void addQuery(int id) {
		synchronized (this._queryIds) {
			this._queryIds.add(id);
			this._dataQueue.add(new MyData(id, "begin"));
		}
	}
	
	protected void removeQuery(int id) {
		synchronized (this._queryIds) {
			this._queryIds.remove(id);
		}
	}
	
	protected void close() {
		synchronized (this._queryIds) {
			this._isClosed.set(true);
			this._dataQueue.clear();
			this._dataQueue.add(null);
		}
	}
	
	private void feedData() {
		
		while (!this._isClosed.get()) {
			
			try {
				Thread.sleep(500);
			} catch (InterruptedException e) { }

            synchronized (this._queryIds) {
                for (int id : this._queryIds) {
                    this._dataQueue.add(new MyData(id, Double.toString(Math.random())));
                }
            }
        }
	}
	
	public MyData receive() {
		
		if (this._isClosed.get()) return null;

        if (Math.random() < 0.01) {
            throw new MyClientException("Error occurred when receive.");
        }

        try {
			return this._dataQueue.take();
		} catch (InterruptedException e) {
			return null;
		}
	}
}
